### ISSUE TITLE
Mapping a Dataset squashes class_names from image_dataset_from_directory

### ISSUE THREAD

**Issue by jamesbraza** [keras-team/tf-keras#592]
**System information**.

- Have I written custom code (as opposed to using a stock example script provided in Keras): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Big Sur v11.5.1
- TensorFlow installed from (source or binary): binary (`pip`)
- TensorFlow version (use command below):2.4.4
- Python version: 3.8.13
- Bazel version (if compiling from source): n/a
- GPU model and memory: n/a
- Exact command to reproduce: see below

**Describe the current behavior**.

I am coming from https://github.com/tensorflow/tensorflow/issues/56233.

Using [image_dataset_from_directory](https://www.tensorflow.org/api_docs/python/tf/keras/utils/image_dataset_from_directory) will correctly create a `Dataset` that contains the labels as `class_names`.

However, when you transform that dataset (e.g. via .map) the `class_names` get clobbered.

**Describe the expected behavior**.

Ideally, `.map`'ing or `.batch`'ing a `Dataset` won't clobber `class_names`.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):

**Standalone code to reproduce the issue**.

From [Image classification from scratch](https://keras.io/examples/vision/image_classification_from_scratch/)

```shell
curl -O https://download.microsoft.com/download/3/E/1/3E1C3F21-ECDB-4869-8368-6DEBA77B919F/kagglecatsanddogs_5340.zip
unzip kagglecatsanddogs_5340.zip
```

Then running this:

```python
import tensorflow as tf

train_ds = tf.keras.preprocessing.image_dataset_from_directory(
    "PetImages",
    batch_size=32,
)
assert hasattr(train_ds, "class_names")  # Passes
mapped_train_ds = train_ds.map(lambda x, y: (x, y))
assert hasattr(mapped_train_ds, "class_names")  # Doesn't pass
```

**Source code / logs**.

n/a
